### PR TITLE
Add a Coordinape User in all circles

### PR DESCRIPTION
**Pull request by crabsinger**
# Summary

Add a Coordinape User to all circles, with a fixed address - to allow the protocol to receive gifts from users.

Goes with FE: https://github.com/coordinape/coordinape/pull/235

https://linear.app/yearn/issue/APE-305/create-a-coordinape-user

```
> docker exec -it app php artisan ape:populate_coordinate_user
```

Based off the staging branch - because that's what contains the new Circle Create API: https://github.com/coordinape/coordinape-backend/pull/47

Questions:

- so it looks like we should be creating one (1) new `profile` record for the Coordinape user, and many (n) new `user`s  records for this user in each circle. This seems like it will duplicate data but I think that's ok for now

## Deploy Instructions

1. Merge FE PR
2. Merge this PR
3. populate COORDINAPE_USER_ADDRESS env on server
4. Run `php artisan ape:populate_coordinape_user`

### CODE REVIEW

File: app/Console/Commands/AddCoordinapeUserToAllCircles.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $address = strtolower($this->argument('address'));
+
+        $profile = Profile::firstOrCreate([
+            'address' => $address,
+        ]);
+
+        # Iterate across all circles (TODO: can we do this in one transaction?)
+        Circle::chunk(100, function ($circles) {
+            foreach ($circles as $circle) {
```
Comment (crabsinger) [inline app/Console/Commands/AddCoordinapeUserToAllCircles.php:51]: @zashton silly question: what's the best way to pass $address into the context of the callback function here?

https://laravel.com/docs/8.x/collections#method-each

I'm getting: `  Undefined variable $address`

Comment (zashton) [reply app/Console/Commands/AddCoordinapeUserToAllCircles.php:51]: nice, just added in for you. you can add a new user type for this as well if you want

**Comment by crabsinger** [comment]
@zashton so any ideas on best way to lock down this new user? Stop anyone including admins from changing its properties? We would only want to let admins remove/re-add the user to a circle?

**Comment by zashton** [comment]
@crabsinger you can disable/hide this user on the FE admin dashboard and then on the backend block this user type from being updated. We do not have a super admin dashboard, so if you need to you can soft delete the user from the db.

File: app/Models/User.php [hunk 2]
```diff
@@ -154,4 +154,8 @@ public function profile() {
     public function nominations() {
         return $this->belongsToMany('App\Models\Nominee','vouches','voucher_id','nominee_id');
     }
+
+    public function isCoordinapeUser() {
```
Comment (zashton) [inline app/Models/User.php:158]: `public function getIsCoordinapeUserAttribute() {
    return $this->role == config('enums.user_types.coordinape');
}`

you could just do this, if you are using role to track whether its coordinape user

File: app/Http/Controllers/UserController.php [hunk 3]
```diff
@@ -55,6 +55,10 @@ public function adminUpdateUser(AdminUserRequest $request, $circle_id, $address)
         $user = $request->user;
         if(!$user)
             return response()->json(['message'=> 'Address not found'],422);
+
+        if($user->isCoordinapeUser == True) #TODO
+            return response()->json(['message'=> 'This user is not modifyable'],423);
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Http/Controllers/UserController.php:60]: Yeap this works to block of the editing the user
`if ($user->isCoordinapeUser)`

File: app/Repositories/CircleRepository.php [hunk 4]
```diff
@@ -34,9 +34,26 @@ public function createCircle($request) {
 
         $address = strtolower($data['address']);
         $circle = $this->model->create(['name' => $data['circle_name'], 'protocol_id' => $protocol_id]);
```
[the remark this answers is not in the document]
Comment (crabsinger) [reply app/Repositories/CircleRepository.php]: ok, added a DB::transaction.

In localhost, I see the error just as a SQL error. @zashton will it show this SQL messsage in the error in prod or wrap it in another error?

File: app/Repositories/CircleRepository.php [hunk 5]
```diff
@@ -34,9 +34,26 @@ public function createCircle($request) {
...
+
         $user = new User(['name' => $data['user_name'], 'circle_id' => $circle->id,
             'role' => config('enums.user_types.admin'), 'address' => $address]);
         $user->save();
+
+        // TODO: move to function and re-use
+        $coordinape_user = new User([
+                    'address' => env('COORDINAPE_USER_ADDRESS'),
```
[the remark this answers is not in the document]
Comment (crabsinger) [reply app/Repositories/CircleRepository.php]: ok, created an config/ape.php file.

**Comment by crabsinger** [comment]
@zashton PTAL. I've updated most of the outstanding points. One remaining uqestion on the IsCoordinapeUser accessor, as it was erroring for me.

Let me know what else you see before this is ready for merge.

**Comment by zashton** [comment]
> @zashton PTAL. I've updated most of the outstanding points. One remaining uqestion on the IsCoordinapeUser accessor, as it was erroring for me.
> 
> Let me know what else you see before this is ready for merge.

i assume you are using it like `$user->isCoordinapeUser()`  ? instead use it like `$user->isCoordinapeUser`

File: app/Repositories/CircleRepository.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (zashton) [reply app/Repositories/CircleRepository.php]: would it use address from the console command or from env ? might need to remove the address from the console command if that is so. 

Looking ahead there might be a need to have an easy way to update the address in future

**Comment by crabsinger** [comment]
@zashton, pushed the changes to the accessor method. Looks like its working now. 👍 

If we want to update the Coordinape User address right now we'd need to do:
1. change ENV variable (this will make new users have this new address)
2. update all old users's addresses to the new address.

I guess we could create a helper method to update all the old addresses... or just do it SQL? Thoughts?

File: app/Console/Commands/AddCoordinapeUserToAllCircles.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    public function handle()
+    {
+        $address = config('ape.coordinape_user_address');
+
+        $profile = Profile::firstOrCreate([
+            'address' => $address,
+        ]);
+
+
+        # Iterate across all circles (TODO: can we do this in one transaction?)
```
Comment (zashton) [inline app/Console/Commands/AddCoordinapeUserToAllCircles.php:52]: yeah. its possible to do it with DB::insert($array)

File: config/ape.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    |--------------------------------------------------------------------------
+    | coordinape_user_address
+    |--------------------------------------------------------------------------
+    |
+    | This value is the address of the Coordinape User. This user is added to
+    | every circle by default and receives GIVE during epochs.
+    |
+    */
+
+    'coordinape_user_address' => env('COORDINAPE_USER_ADDRESS'),
```
Comment (zashton) [inline config/ape.php:15]: could we replace `DEV_ETH_ADDRESS` in env.example with this

File: app/Models/Uxresearch.php [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace App\Models;
```
Comment (zashton) [inline app/Models/Uxresearch.php:3]: this file can be deleted , no longer used anywhere

**Review by zashton — CHANGES_REQUESTED** [review]

File: app/Console/Commands/AddCoordinapeUserToAllCircles.php (hunk 6 above)
Comment (levity) [reply app/Console/Commands/AddCoordinapeUserToAllCircles.php:52]: or wrap in DB::transaction?

File: app/Repositories/CircleRepository.php [hunk 9]
```diff
@@ -85,4 +90,20 @@ public function getWebhook($circle_id) {
         $circle = $this->model->find($circle_id);
        return $circle->discord_webhook ?:'';
     }
+
+    public function addCoordinapeUserToCircle($circle_id) {
+        $coordinape_user = new User([
+                    'address' => config('ape.coordinape_user_address'),
```
Comment (levity) [inline app/Repositories/CircleRepository.php:96]: indentation seems off here?

**Review by zashton — CHANGES_REQUESTED** [review]

File: app/Repositories/CircleRepository.php [hunk 10]
```diff
@@ -85,4 +90,20 @@ public function getWebhook($circle_id) {
         $circle = $this->model->find($circle_id);
        return $circle->discord_webhook ?:'';
     }
+
+    public function addCoordinapeUserToCircle($circle_id) {
```
Comment (levity) [inline app/Repositories/CircleRepository.php:94]: maybe you could change this to take an array of circle IDs and then bulk insert as Zashton suggested above

Comment (zashton) [reply app/Repositories/CircleRepository.php:94]: yeah if you're doing that
move 
`$profile = Profile::firstOrCreate([ 'address' => config('ape.coordinape_user_address') ]);`
this inside the function as well

File: app/Repositories/CircleRepository.php [hunk 11]
```diff
@@ -24,28 +25,32 @@ public function getCircles($request) {
...
+            } else {
+                $protocol_id = $data['protocol_id'];
+            }
+
+            $address = strtolower($data['address']);
+            $circle = $this->model->create(['name' => $data['circle_name'], 'protocol_id' => $protocol_id]);
+            $user = new User(['name' => $data['user_name'], 'circle_id' => $circle->id,
+                'role' => config('enums.user_types.admin'), 'address' => $address]);
+            $user->save();
+            self::addCoordinapeUserToCircle($circle->id);
```
[the remark this answers is not in the document]
Comment (crabsinger) [reply app/Repositories/CircleRepository.php:43]: ah, good catch! fixed.

File: config/ape.php (hunk 7 above)
Comment (crabsinger) [reply config/ape.php:15]: done

File: app/Repositories/CircleRepository.php (hunk 9 above)
Comment (crabsinger) [reply app/Repositories/CircleRepository.php]: linted and pushed.

**Review by levity — APPROVED** [review]

File: app/Console/Commands/AddCoordinapeUserToAllCircles.php (hunk 6 above)
Comment (crabsinger) [reply app/Console/Commands/AddCoordinapeUserToAllCircles.php:52]: lets do this later, i think its not needed for now.

File: app/Repositories/CircleRepository.php (hunk 10 above)
Comment (crabsinger) [reply app/Repositories/CircleRepository.php]: that would make this better, but lets merge this version first.

**Comment by levity** [comment]
:shipit:

**Pull request merged**